cin_validator/rules/cin2022_23/rule_8839.py

```diff
@@ -0,0 +1,214 @@
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[DateOfInitialCPC],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # PREPARING DATA
+
+    # Replace Assessments with the name of the table you need.
+    df = data_container[Section47]
+    # Before you begin, rename the index and make it a column, so that the initial row positions can be kept intact.
+    df.index.name = "ROW_ID"
+    df.reset_index(inplace=True)
+
+    # lOGIC
+    # Implement rule logic as described by the Github issue.
+    # Put the description as a comment above the implementation as shown.
+
+    # Within one <CINdetails> group, there must be only one <Section47> group that has no <DateOfInitialCPC> (N00110) recorded
+    # OR <Section47> group that has a missing <DateOfInitialCPC> (N00110) and the <ICPCnotRequired> (N00111) flag is not true
+
+    # DF_CHECK: APPLY GROUPBYs IN A SEPARATE DATAFRAME SO THAT OTHER COLUMNS ARE NOT LOST OR CORRUPTED. THEN, MAP THE RESULTS TO THE INITIAL DATAFRAME.
+    df_check = df.copy()
+    # get all the locations where ICPCnotRequired is null or not true (1)
+    df_check = df_check[
+        df_check[ICPCnotRequired].isna() | (df_check[ICPCnotRequired] != 1)
+    ]
+    # get all the locations where DateOfInitialCPC is null
+    df_check = df_check[df_check[DateOfInitialCPC].isna()]
+    # We'll have to count the number of nan values per group. NaNs cannot be counted so replace them with something that can.
+    # Do this only if your rule requires that you interact with a column made up of all NaNs.
+    # Adding as a separate column because I need the original column in the groupby for df_issues later on.
+    df_check["CountICPC"] = df_check[DateOfInitialCPC]
+    df_check["CountICPC"].fillna(1, inplace=True)
+    # count how many occurences of missing DateOfInitialCPC per CINdetails group in each child.
+    df_check = (
+        df_check.groupby([LAchildID, CINdetailsID, DateOfInitialCPC], dropna=False)[
+            "CountICPC"
+        ]
+        .count()
+        .reset_index()
+    )
+    #
+    # when you groupby as shown above a series is returned where the columns in the round brackets become the index and the groupby result are the values.
+    # resetting the index pushes the columns in the () back as columns of the dataframe and assigns the groupby result to the column in the square bracket.
+    #
+    # filter out the instances where DateOfInitialCPC is missing more than once in a CINdetails group.
+    df_check = df_check[df_check["CountICPC"] > 1]
+    issue_ids = tuple(
+        zip(df_check[LAchildID], df_check[CINdetailsID], df_check[DateOfInitialCPC])
+    )
+    # DF_ISSUES: GET ALL THE DATA ABOUT THE LOCATIONS THAT WERE IDENTIFIED IN DF_CHECK
+    df["ERROR_ID"] = tuple(zip(df[LAchildID], df[CINdetailsID], df[DateOfInitialCPC]))
+    df_issues = df[df.ERROR_ID.isin(issue_ids)]
```

Comment:
Note I've deviated from the example in #172 because I found it doesn't work when there are a mixture of open and closed records (Section 47s here, Assessments in 8896) within the same CINdetailsID group.  I have to include the null date in the error ID to avoid flagging those incorrectly in the error table.

Reply:
So for cases where it is necessary to map issue_ids from a filtered df (df_check in this case) to the original dataframe (df in this case), we can say that it is okay to leave in the redundant columns of ERROR_ID since they might act as a differentiating factor during the df matching.